donkeycar/pipeline/sequence.py

```diff
@@ -0,0 +1,217 @@
+    def __init__(self,
+                 iterator: Iterable[R],
+                 x_transform: Callable[[R], XOut],
+                 y_transform: Callable[[R], YOut]) -> None:
+
+        self.iterator = BaseTfmIterator_(
+            iterator=iterator, x_transform=x_transform, y_transform=y_transform)
+
+    def __len__(self):
+        return len(self.iterator)
+
+    def __iter__(self) -> SizedIterator[Tuple[XOut, YOut]]:
+        return self
+
+    def __next__(self):
+        return next(self.iterator)
+
+    def batched(self, batch_size=64) -> SizedIterator[List[Tuple[XOut, YOut]]]:
+        return self.iterator.batched(batch_size=batch_size)
+
+
+class TfmTupleIterator(Generic[X, Y, XOut, YOut],  SizedIterator[Tuple[XOut, YOut]]):
+    def __init__(self,
+                 iterator: Iterable[Tuple[X, Y]],
+                 x_transform: Callable[[X], XOut],
+                 y_transform: Callable[[Y], YOut]) -> None:
+
+        self.iterator = BaseTfmIterator_(
+            iterator=iterator, x_transform=x_transform, y_transform=y_transform)
+
+    def __len__(self):
+        return len(self.iterator)
+
+    def __iter__(self) -> SizedIterator[Tuple[XOut, YOut]]:
+        return self
+
+    def __next__(self):
+        return next(self.iterator)
+
+    def batched(self, batch_size=64) -> SizedIterator[List[Tuple[XOut, YOut]]]:
+        return self.iterator.batched(batch_size=batch_size)
+
+
+class BaseTfmIterator_(Generic[XOut, YOut],  SizedIterator[Tuple[XOut, YOut]]):
+    '''
+    A basic transforming iterator.
+    Do no use this class directly.
+    '''
+
+    def __init__(self,
+                 # Losing a bit of type safety here for a common implementation
+                 iterator: Iterable[Any],
+                 x_transform: Callable[[R], XOut],
+                 y_transform: Callable[[R], YOut]) -> None:
+
+        self.iterator = iter(iterator)
+        self.x_transform = x_transform
+        self.y_transform = y_transform
+
+    def __len__(self):
+        return len(self.iterator)
+
+    def __iter__(self) -> SizedIterator[Tuple[XOut, YOut]]:
+        return self
+
+    def __next__(self):
+        record = next(self.iterator)
+        if (isinstance(record, tuple) and len(record) == 2):
+            x, y = record
+            return self.x_transform(x), self.y_transform(y)
+        else:
+            return self.x_transform(record), self.y_transform(record)
+
+    def batched(self, batch_size=64) -> SizedIterator[List[Tuple[XOut, YOut]]]:
+        batched_iterator = BatchedIterator(self, batch_size=batch_size)
+        return batched_iterator
+
+
+class BatchedIterator(SizedIterator[List[R]]):
+    def __init__(self, pipeline: SizedIterator[R], batch_size: int) -> None:
+        self.pipeline = pipeline
+        self.batch_size = batch_size
+        self.consumed = 0
+
+    def __len__(self):
+        return math.ceil(len(self.pipeline) / self.batch_size)
```

Comment:
Good to use `ceil` because before we were neglecting the the last non-completely filled batch.